src/include/btmem.h

```diff
@@ -1600,23 +1600,35 @@ struct __wt_col_fix_auxiliary_header {
  * examining an index, we don't want the oldest split generation to move forward and potentially
  * free it.
  */
-#define WT_ENTER_PAGE_INDEX(session)                                         \
-    do {                                                                     \
-        uint64_t __prev_split_gen = __wt_session_gen(session, WT_GEN_SPLIT); \
-        if (__prev_split_gen == 0)                                           \
-            __wt_session_gen_enter(session, WT_GEN_SPLIT);
-
-#define WT_LEAVE_PAGE_INDEX(session)                   \
-    if (__prev_split_gen == 0)                         \
-        __wt_session_gen_leave(session, WT_GEN_SPLIT); \
-    }                                                  \
+#define WT_ENTER_PAGE_INDEX(session)      \
+    do {                                  \
+        bool __entered_split_gen = false; \
+        WT_ENTER_GENERATION((session), WT_GEN_SPLIT, __entered_split_gen);
+
+#define WT_LEAVE_PAGE_INDEX(session)                                   \
+    WT_LEAVE_GENERATION((session), WT_GEN_SPLIT, __entered_split_gen); \
+    }                                                                  \
     while (0)
 
 #define WT_WITH_PAGE_INDEX(session, e) \
     WT_ENTER_PAGE_INDEX(session);      \
     (e);                               \
     WT_LEAVE_PAGE_INDEX(session)
 
+/*
+ * Manage the given generation number with support for re-entry. Re-entry is allowed as the previous
+ * generation as it must be as low as the current generation.
+ */
+#define WT_ENTER_GENERATION(session, generation, entered) \
```

Comment:
Could this do something like:

```
bool __entered_##generation = false;
```

That would mean that callers don't need to create a local variable. It would also require wrapping this in a do-while, and so users need to be careful to acquire in one order and release in the other. So it might not be worth the complexity.

Maybe it would be better to move more towards all ENTER/LEAVE usage moving to a `WT_WITH` pattern, similar to `WT_WITH_PAGE_INDEX`. That's neater unless error handling gets in the way.

Reply:
For the case that this is targeted the WT_WITH is not a great pattern due to the error handling. I think the __entered_#generation is a great idea, even with the do-while usage. It makes it much cleaner. Pushed. Please review.